crates/bevy_reflect/src/set.rs

```diff
@@ -0,0 +1,530 @@
+    }
+
+    /// Inserts a typed value into the set.
+    pub fn insert<V: Reflect>(&mut self, value: V) {
+        self.insert_boxed(Box::new(value));
+    }
+
+    #[allow(clippy::borrowed_box)]
+    fn internal_hash(value: &Box<dyn Reflect>) -> u64 {
+        value.reflect_hash().expect(hash_error!(value))
+    }
+
+    #[allow(clippy::borrowed_box)]
+    fn internal_eq(value: &Box<dyn Reflect>) -> impl FnMut(&Box<dyn Reflect>) -> bool + '_ {
+        |other| {
+            value
+                .reflect_partial_eq(&**other)
+                .expect("Not the same type")
+        }
+    }
+}
+
+// I just created this function to have only one point where we ignore the rust warning about the
+// unused allocation
+fn box_and_clone(val: &dyn Reflect) -> Box<dyn Reflect> {
+    #[allow(unused_allocation)]
+    Box::new(val).clone_value()
+}
+
+impl Set for DynamicSet {
+    fn get(&self, value: &dyn Reflect) -> Option<&dyn Reflect> {
+        let boxed = box_and_clone(value);
+        self.hash_table
+            .find(Self::internal_hash(&boxed), Self::internal_eq(&boxed))
+            .map(|value| &**value)
+    }
+
+    fn len(&self) -> usize {
+        self.hash_table.len()
+    }
+
+    fn iter(&self) -> Box<dyn Iterator<Item = &dyn Reflect> + '_> {
+        let iter = self.hash_table.iter().map(|v| &**v);
+        Box::new(iter)
+    }
+
+    fn drain(self: Box<Self>) -> Vec<Box<dyn Reflect>> {
+        self.hash_table.into_iter().collect::<Vec<_>>()
+    }
+
+    fn clone_dynamic(&self) -> DynamicSet {
+        let mut hash_table = HashTable::new();
+        self.hash_table
+            .iter()
+            .map(|value| value.clone_value())
+            .for_each(|value| {
+                hash_table.insert_unique(Self::internal_hash(&value), value, Self::internal_hash);
+            });
+
+        DynamicSet {
+            represented_type: self.represented_type,
+            hash_table,
+        }
+    }
+
+    fn insert_boxed(&mut self, value: Box<dyn Reflect>) -> bool {
+        match self
+            .hash_table
+            .find(Self::internal_hash(&value), Self::internal_eq(&value))
+        {
+            Some(_) => {
+                self.hash_table.insert_unique(
+                    Self::internal_hash(&value),
+                    value,
+                    Self::internal_hash,
+                );
+                false
+            }
+            None => {
+                self.hash_table.insert_unique(
+                    Self::internal_hash(&value),
+                    value,
+                    Self::internal_hash,
+                );
+                true
+            }
+        }
+    }
+
+    fn remove(&mut self, value: &dyn Reflect) -> bool {
+        let prev_len = self.hash_table.len();
+        let boxed = box_and_clone(value);
+        self.hash_table.retain(|v| !Self::internal_eq(&boxed)(v));
+        let post_len = self.hash_table.len();
+        prev_len != post_len
+    }
```

Comment:
You can remove specific entries in a `HashTable` by using `find_entry` and calling `remove` on the `OccupiedEntry` you get.